File: connectors/javascript/tests/integration/advanced/A04_OnErrorEvent.test.ts
/**
 * A-04: OnError Event Tests
 *
 * Tests for the OnError event which fires on request failures,
 * connection problems, server errors, and other error conditions.
 * Used for error handling in callback-based API.
 */

import { describe, test, expect, beforeEach, afterEach } from 'vitest';
import { BaseIntegrationTest } from '../helpers/BaseIntegrationTest.js';
import { Packet } from '../../../src/packet.js';

describe('A-04: OnError Event', () => {
    let testContext: BaseIntegrationTest;

    beforeEach(async () => {
        testContext = new BaseIntegrationTest();
        await testContext['beforeEach']();
    });

    afterEach(async () => {
        await testContext['afterEach']();
    });

    test('A-04-01: OnError fires when send() on disconnected state', async () => {
        // Given: Connected, authenticated, then disconnected
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('onerror-user-1');

        const errorEvents: Array<{ stageId: bigint; errorCode: number; request?: Packet }> = [];
        testContext['connector']!.onError = (stageId: bigint, errorCode: number, message: string, request?: Packet) => {
            errorEvents.push({ stageId, errorCode, request });
        };

        testContext['connector']!.disconnect();
        await testContext['delay'](500);

        // When: Try to send after disconnect
        const echoRequest = Packet.empty('EchoRequest');
        testContext['connector']!.send(echoRequest);
        testContext['connector']!.mainThreadAction();

        // Then: OnError should be fired
        expect(errorEvents.length).toBeGreaterThanOrEqual(1);
        expect(errorEvents[0].errorCode).toBeGreaterThan(0);
        expect(errorEvents[0].stageId).toBeDefined();
    });

    test('A-04-02: OnError fires when request() callback on disconnected state', async () => {
        // Given: Connected, authenticated, then disconnected
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('onerror-user-2');

        const errorEvents: Array<{ stageId: bigint; errorCode: number }> = [];
        testContext['connector']!.onError = (stageId: bigint, errorCode: number) => {
            errorEvents.push({ stageId, errorCode });
        };

        testContext['connector']!.disconnect();
        await testContext['delay'](500);

        // When: Try to request after disconnect using callback
        const echoRequest = Packet.empty('EchoRequest');
        let callbackFired = false;

        testContext['connector']!.requestWithCallback(echoRequest, (response: Packet) => {
            callbackFired = true;
        });
        testContext['connector']!.mainThreadAction();

        // Then: OnError should fire, callback should not
        expect(errorEvents.length).toBeGreaterThanOrEqual(1);
        expect(errorEvents[0].errorCode).toBeGreaterThan(0);
        expect(callbackFired).toBe(false);
    });

    test('A-04-03: OnError contains original request packet', async () => {
        // Given: Disconnected state
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('onerror-user-3');

        let receivedRequest: Packet | undefined;
        testContext['connector']!.onError = (stageId: bigint, errorCode: number, message: string, request?: Packet) => {
            receivedRequest = request;
        };

        testContext['connector']!.disconnect();
        await testContext['delay'](500);

        // When: Send request
        const echoRequest = Packet.empty('EchoRequest');
        testContext['connector']!.send(echoRequest);
        testContext['connector']!.mainThreadAction();

        // Then: Original request should be in OnError
        expect(receivedRequest).toBeDefined();
        expect(receivedRequest?.msgId).toBe('EchoRequest');
    });

    test('A-04-04: OnError contains StageId information', async () => {
        // Given: Connected to specific stage
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('onerror-user-4');

        const originalStageId = testContext['connector']!.stageId;

        let receivedStageId: bigint | undefined;
        testContext['connector']!.onError = (stageId: bigint) => {
            receivedStageId = stageId;
        };

        testContext['connector']!.disconnect();
        await testContext['delay'](500);

        // When: Send after disconnect
        const echoRequest = Packet.empty('EchoRequest');
        testContext['connector']!.send(echoRequest);
        testContext['connector']!.mainThreadAction();

        // Then: StageId should match original
        expect(receivedStageId).toBe(originalStageId);
    });

    test('A-04-05: OnError fires when authenticate() on disconnected state', async () => {
        // Given: Connected then disconnected
        await testContext['createStageAndConnect']();

        const errorEvents: number[] = [];
        testContext['connector']!.onError = (stageId: bigint, errorCode: number) => {
            errorEvents.push(errorCode);
        };

        testContext['connector']!.disconnect();
        await testContext['delay'](500);

        // When: Try to authenticate after disconnect using callback
        const authRequest = Packet.empty('AuthenticateRequest');
        let callbackFired = false;

        testContext['connector']!.requestWithCallback(authRequest, (response: Packet) => {
            callbackFired = true;
        });
        testContext['connector']!.mainThreadAction();

        // Then: OnError should fire
        expect(errorEvents.length).toBeGreaterThanOrEqual(1);
        expect(errorEvents[0]).toBeGreaterThan(0);
        expect(callbackFired).toBe(false);
    });

    test('A-04-06: Multiple OnError handlers can be registered', async () => {
        // Given: Connected and authenticated
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('onerror-user-6');

        let handler1Called = false;
        let handler2Called = false;
        let handler3Called = false;

        // Note: JavaScript typically supports only one handler per event
        // This test verifies the last handler is called
        testContext['connector']!.onError = () => { handler1Called = true; };
        testContext['connector']!.onError = () => { handler2Called = true; };
        testContext['connector']!.onError = () => { handler3Called = true; };

        testContext['connector']!.disconnect();
        await testContext['delay'](500);

        // When: Trigger error
        const echoRequest = Packet.empty('EchoRequest');
        testContext['connector']!.send(echoRequest);
        testContext['connector']!.mainThreadAction();

        // Then: At least one handler should be called (likely the last one)
        const anyHandlerCalled = handler1Called || handler2Called || handler3Called;
        expect(anyHandlerCalled).toBe(true);
    });

    test('A-04-07: OnError handler exception does not crash connector', async () => {
        // Given: Connected and authenticated
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('onerror-user-7');

        let handler2Called = false;

        testContext['connector']!.onError = () => {
            handler2Called = true;
            throw new Error('Test exception in handler');
        };

        testContext['connector']!.disconnect();
        await testContext['delay'](500);

        // When: Trigger error
        const echoRequest = Packet.empty('EchoRequest');

        try {
            testContext['connector']!.send(echoRequest);
            testContext['connector']!.mainThreadAction();
        } catch (error) {
            // Exception may or may not propagate depending on implementation
        }

        // Then: Handler was called despite exception
        expect(handler2Called).toBe(true);
    });

    test('A-04-08: Normal disconnect does not fire OnError', async () => {
        // Given: Connected and authenticated
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('onerror-user-8');

        let errorCount = 0;
        testContext['connector']!.onError = () => {
            errorCount++;
        };

        // When: Normal disconnect
        testContext['connector']!.disconnect();
        await testContext['delay'](500);

        // Then: OnError should not fire for normal disconnect
        expect(errorCount).toBe(0);
    });
});

File: connectors/cpp/tests/integration/advanced/a04_on_error_event_test.cpp
#include "../base_integration_test.hpp"
#include <gtest/gtest.h>
#include <atomic>

using namespace playhouse;
using namespace playhouse::test;

/// A-04: OnError Event Tests
/// Verifies OnError event callback mechanism
class A04_OnErrorEventTest : public BaseIntegrationTest {};

TEST_F(A04_OnErrorEventTest, OnError_ConnectionClosed_EventFires) {
    // Given: Connected to server with OnError handler
    ASSERT_TRUE(CreateStageAndConnect());

    std::atomic<bool> error_fired{false};
    int error_code_received = 0;
    std::string error_message;

    connector_->OnError = [&](int code, std::string message) {
        error_fired = true;
        error_code_received = code;
        error_message = message;
    };

    // When: Force connection error by disconnecting and attempting operation
    connector_->Disconnect();
    std::this_thread::sleep_for(std::chrono::milliseconds(500));

    std::string data = "{\"test\":\"data\"}";
    Bytes payload(data.begin(), data.end());
    connector_->Send(Packet::FromBytes("TestMessage", std::move(payload)));

    // Process callbacks
    bool completed = WaitForConditionWithMainThreadAction([&]() {
        return error_fired.load();
    }, 5000);

    // Then: OnError should fire with connection closed error
    EXPECT_TRUE(completed) << "OnError event should fire";
    EXPECT_TRUE(error_fired);
    EXPECT_EQ(error_code_received, error_code::CONNECTION_CLOSED);
    EXPECT_FALSE(error_message.empty());
}

TEST_F(A04_OnErrorEventTest, OnError_MultipleHandlers_CanBeUpdated) {
    // Given: Connected to server
    ASSERT_TRUE(CreateStageAndConnect());

    std::atomic<int> handler1_count{0};
    std::atomic<int> handler2_count{0};

    // First handler
    connector_->OnError = [&](int code, std::string message) {
        handler1_count++;
    };

    // Trigger error
    connector_->Disconnect();
    std::this_thread::sleep_for(std::chrono::milliseconds(200));

    std::string data1 = "{\"test\":\"1\"}";
    Bytes payload1(data1.begin(), data1.end());
    connector_->Send(Packet::FromBytes("Test1", std::move(payload1)));

    // Process callbacks
    for (int i = 0; i < 10; i++) {
        connector_->MainThreadAction();
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }

    // When: Update handler
    connector_->OnError = [&](int code, std::string message) {
        handler2_count++;
    };

    std::string data2 = "{\"test\":\"2\"}";
    Bytes payload2(data2.begin(), data2.end());
    connector_->Send(Packet::FromBytes("Test2", std::move(payload2)));

    // Process callbacks
    for (int i = 0; i < 10; i++) {
        connector_->MainThreadAction();
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }

    // Then: Second handler should fire, not first
    EXPECT_EQ(handler1_count.load(), 1) << "First handler should fire once";
    EXPECT_EQ(handler2_count.load(), 1) << "Second handler should fire once";
}

TEST_F(A04_OnErrorEventTest, OnError_WithoutHandler_DoesNotCrash) {
    // Given: Connected without OnError handler
    ASSERT_TRUE(CreateStageAndConnect());

    // OnError should be cleared for this test
    connector_->OnError = nullptr;
    EXPECT_FALSE(connector_->OnError);

    // When: Trigger error condition
    connector_->Disconnect();
    std::this_thread::sleep_for(std::chrono::milliseconds(200));

    std::string data = "{\"test\":\"data\"}";
    Bytes payload(data.begin(), data.end());
    connector_->Send(Packet::FromBytes("TestMessage", std::move(payload)));

    // Process callbacks
    for (int i = 0; i < 10; i++) {
        connector_->MainThreadAction();
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }

    // Then: Should not crash even without handler
    SUCCEED() << "Error handled without OnError handler";
}

TEST_F(A04_OnErrorEventTest, OnError_ErrorCodeTypes_Differentiated) {
    // Given: Connected to server
    ASSERT_TRUE(CreateStageAndConnect());

    std::vector<int> error_codes;

    connector_->OnError = [&](int code, std::string message) {
        error_codes.push_back(code);
    };

    // When: Trigger connection closed error
    connector_->Disconnect();
    std::this_thread::sleep_for(std::chrono::milliseconds(200));

    std::string data = "{\"test\":\"data\"}";
    Bytes payload(data.begin(), data.end());
    connector_->Send(Packet::FromBytes("TestMessage", std::move(payload)));

    // Process callbacks
    for (int i = 0; i < 10; i++) {
        connector_->MainThreadAction();
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }

    // Then: Should receive CONNECTION_CLOSED error code
    EXPECT_FALSE(error_codes.empty()) << "Should receive at least one error code";
    if (!error_codes.empty()) {
        EXPECT_EQ(error_codes[0], error_code::CONNECTION_CLOSED);
    }
}

TEST_F(A04_OnErrorEventTest, OnError_MessageProvided_NotEmpty) {
    // Given: Connected to server
    ASSERT_TRUE(CreateStageAndConnect());

    std::atomic<bool> message_not_empty{false};

    connector_->OnError = [&](int code, std::string message) {
        message_not_empty = !message.empty();
    };

    // When: Trigger error
    connector_->Disconnect();
    std::this_thread::sleep_for(std::chrono::milliseconds(200));

    std::string data = "{\"test\":\"data\"}";
    Bytes payload(data.begin(), data.end());
    connector_->Send(Packet::FromBytes("TestMessage", std::move(payload)));

    // Process callbacks
    bool completed = WaitForConditionWithMainThreadAction([&]() {
        return message_not_empty.load();
    }, 5000);

    // Then: Error message should be provided
    EXPECT_TRUE(completed);
    EXPECT_TRUE(message_not_empty) << "Error message should not be empty";
}
